pkg/apiserver/authorization/authorizerfactory/opa.go

```diff
@@ -33,125 +34,140 @@ type opaAuthorizer struct {
+func (o *opaAuthorizer) makeDecision(role interface{}, a authorizer.Attributes) (authorized authorizer.Decision, reason string, err error) {
 
-	for _, ruleRef := range role.Rules {
-		rule, err := o.am.GetPolicyRule(ruleRef.Name)
-		if err != nil {
-			if errors.IsNotFound(err) {
-				continue
-			}
-			return authorizer.DecisionDeny, "", err
-		}
-		// Call the rego.New function to create an object that can be prepared or evaluated
-		//  After constructing a new rego.Rego object you can call PrepareForEval() to obtain an executable query
-		query, err := rego.New(rego.Query(defaultRegoQuery), rego.Module("authz.rego", rule.Rego)).PrepareForEval(context.Background())
+	regoPolicy := ""
 
-		if err != nil {
-			klog.Errorf("rule syntax error:%s", err)
-			continue
+	// override
+	if globalRole, ok := role.(*iamv1alpha2.GlobalRole); ok {
+		if overrideRego, ok := globalRole.Annotations[iamv1alpha2.RegoOverrideAnnotation]; ok {
+			regoPolicy = overrideRego
+		}
+	} else if workspaceRole, ok := role.(*iamv1alpha2.WorkspaceRole); ok {
+		if overrideRego, ok := workspaceRole.Annotations[iamv1alpha2.RegoOverrideAnnotation]; ok {
+			regoPolicy = overrideRego
+		}
+	} else if clusterRole, ok := role.(*rbacv1.ClusterRole); ok {
+		if overrideRego, ok := clusterRole.Annotations[iamv1alpha2.RegoOverrideAnnotation]; ok {
+			regoPolicy = overrideRego
+		}
+	} else if role, ok := role.(*rbacv1.Role); ok {
+		if overrideRego, ok := role.Annotations[iamv1alpha2.RegoOverrideAnnotation]; ok {
+			regoPolicy = overrideRego
 		}
+	}
 
-		// The policy decision is contained in the results returned by the Eval() call. You can inspect the decision and handle it accordingly.
-		results, err := query.Eval(context.Background(), rego.EvalInput(a))
+	if regoPolicy == "" {
+		return authorizer.DecisionNoOpinion, "", nil
+	}
 
-		if err != nil {
-			klog.Errorf("rule syntax error:%s", err)
-			continue
-		}
+	// Call the rego.New function to create an object that can be prepared or evaluated
+	//  After constructing a new rego.Rego object you can call PrepareForEval() to obtain an executable query
+	query, err := rego.New(rego.Query(defaultRegoQuery), rego.Module("authz.rego", regoPolicy)).PrepareForEval(context.Background())
 
-		if len(results) > 0 && results[0].Expressions[0].Value == true {
-			return authorizer.DecisionAllow, "", nil
-		}
+	if err != nil {
+		klog.Errorf("syntax error:%s,refer: %s+v", err, role)
+		return authorizer.DecisionNoOpinion, "", err
+	}
+
+	// The policy decision is contained in the results returned by the Eval() call. You can inspect the decision and handle it accordingly.
+	results, err := query.Eval(context.Background(), rego.EvalInput(a))
+
+	if err != nil {
+		klog.Errorf("syntax error:%s,refer: %s+v", err, role)
+		return authorizer.DecisionNoOpinion, "", err
+	}
+
+	if len(results) > 0 && results[0].Expressions[0].Value == true {
```

Comment:
Don't we need to check `len(results[0].Expressions) > 0` first?

Reply:
It's not needed, already checked syntax error.